### PR TITLE
CLOUDSTACK-8609. [VMware] VM is not accessible after a migration acro…

### PR DESCRIPTION
**Pull request by likitha**
…ss clusters.

Once a VM is successfully started, don't delete the files associated with the unregistered VM, if the files are in a storage that is being used by the new VM.
Attempt to unregister a VM in another DC, only if there is a host associated with a VM.

### CODE REVIEW

**Comment by remibergsma** [comment]
Who wants to step in and finish this work? It seems the original author is not able to finish it. If no one steps in, we'll have to close the PR without merging it so please help :-).

File: plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 1]
```diff
@@ -1765,7 +1767,18 @@ protected StartAnswer execute(StartCommand cmd) {
...
+                    vmDatastoreNames.add(vmDatastore.getName());
+                }
+                // Don't delete files that are in a datastore that is being used by the new VM as well (zone-wide datastore).
+                for (DatastoreMO existingDatastore : existingDatastores) {
+                    if (vmDatastoreNames.contains(existingDatastore.getName())) {
+                        skipDatastores.add(existingDatastore.getName());
+                    }
+                }
+                deleteUnregisteredVmFiles(existingVmFileLayout, dcMo, true, skipDatastores);
             }
```
Comment (pedro-martins) [inline plugins/hypervisors/vmware/src/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: Hi, likitha.
Could you extract the code between lines 1770 - 1775 and 1776 - 1780 to methods with a little test case and Javadoc explaining what the methods do, the params that they use and what they return?

Another thing, how about you replace the logic in 'if's at lines 2272, 2283 and 2291 for a method with a little test case and a Javadoc, that receives 2 params, (List<String> skipDatastores, String name ) and returns  skipDatastores == null || !skipDatastores.contains(name).

Ty.

**Comment by rodrigo93** [comment]
Could someone put this PR to test again? If it fails, it could be closed since the author seems to be inactive on github since September of the last year and the PR is very old.
If someone shows up some interest in continuing it, then it could be opened again.

_PS: This is just a suggestion._

File: vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java [hunk 2]
```diff
@@ -880,6 +880,38 @@ else if (prop.getName().startsWith("value[")) {
...
+        PropertyFilterSpec pfSpec = new PropertyFilterSpec();
+        pfSpec.getPropSet().add(pSpec);
+        pfSpec.getObjectSet().add(oSpec);
+        List<PropertyFilterSpec> pfSpecArr = new ArrayList<PropertyFilterSpec>();
+        pfSpecArr.add(pfSpec);
+
+        List<ObjectContent> ocs = _context.getService().retrieveProperties(_context.getPropertyCollector(), pfSpecArr);
+
+        List<DatastoreMO> datastores = new ArrayList<DatastoreMO>();
+        if (ocs != null && ocs.size() > 0) {
```
Comment (GabrielBrascher) [inline vmware-base/src/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java]: Hi @likitha, you could use CollectionUtils.isEmpty for this conditional.
It returns true if the List is empty or null.
Thanks.
Documentation: https://commons.apache.org/proper/commons-collections/apidocs/org/apache/commons/collections4/CollectionUtils.html#isEmpty%28java.util.Collection%29

**Comment by rhtyd** [comment]
tag:vmware-pickup

**Comment by bvbharatk** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 137
 Hypervisor xenserver
 NetworkType Advanced
 Passed=73
 Failed=0
 Skipped=3

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**

**Skipped tests:**
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_vpc_vpn.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_routers.py
test_reset_vm_on_reboot.py
test_snapshots.py
test_deploy_vms_with_varied_deploymentplanners.py
test_login.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_disk_offerings.py

**Comment by sureshanaparti** [comment]
Continued with the code changes here in a new PR #2091.

**Comment by DaanHoogland** [comment]
closing as @sureshanaparti tookk this to #2091

**Pull request closed without merging**